src/libraries/Common/src/System/Reflection/AssemblyNameParser.cs

```diff
@@ -0,0 +1,568 @@
+            {
+                result = Array.Empty<byte>();
+                return true;
+            }
+
+            if (isToken && attributeValue.Length != 8 * 2)
+            {
+                return false;
+            }
+
+            byte[] pkt = new byte[attributeValue.Length / 2];
+            int srcIndex = 0;
+            for (int i = 0; i < pkt.Length; i++)
+            {
+                char hi = attributeValue[srcIndex++];
+                char lo = attributeValue[srcIndex++];
+
+                if (!TryParseHexNybble(hi, out byte parsedHi) || !TryParseHexNybble(lo, out byte parsedLo))
+                {
+                    return false;
+                }
+
+                pkt[i] = (byte)((parsedHi << 4) | parsedLo);
+            }
+            result = pkt;
+            return true;
+        }
+
+        private static bool TryParseProcessorArchitecture(string attributeValue, out ProcessorArchitecture result)
+        {
+            result = attributeValue switch
+            {
+                _ when attributeValue.Equals("msil", StringComparison.OrdinalIgnoreCase) => ProcessorArchitecture.MSIL,
+                _ when attributeValue.Equals("x86", StringComparison.OrdinalIgnoreCase) => ProcessorArchitecture.X86,
+                _ when attributeValue.Equals("ia64", StringComparison.OrdinalIgnoreCase) => ProcessorArchitecture.IA64,
+                _ when attributeValue.Equals("amd64", StringComparison.OrdinalIgnoreCase) => ProcessorArchitecture.Amd64,
+                _ when attributeValue.Equals("arm", StringComparison.OrdinalIgnoreCase) => ProcessorArchitecture.Arm,
+                _ when attributeValue.Equals("msil", StringComparison.OrdinalIgnoreCase) => ProcessorArchitecture.MSIL,
+                _ => ProcessorArchitecture.None
+            };
+            return result != ProcessorArchitecture.None;
+        }
+
+        private static bool TryParseHexNybble(char c, out byte parsed)
+        {
+            int value = HexConverter.FromChar(c);
+            if (value == 0xFF)
+            {
+                parsed = 0;
+                return false;
+            }
+            parsed = (byte)value;
+            return true;
+        }
+
+        private static bool IsWhiteSpace(char ch)
+        {
+            switch (ch)
+            {
+                case '\n':
+                case '\r':
+                case ' ':
+                case '\t':
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool TryGetNextChar(out char ch)
+        {
+            if (_index < _input.Length)
+            {
+                ch = _input[_index++];
+                if (ch == '\0')
+                {
+                    return false;
+                }
+            }
+            else
+            {
+                ch = '\0';
```

Comment:
We return true if _index >= _input.Length? Is that then relying on callers to not invoke this again if it returns true with ch == '\0'?

Reply:
From what I can see the caller maps it to `Token.End`

https://github.com/dotnet/runtime/blob/a2ce8db5fb14df85cef03880a30133be198626e1/src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameParser.cs#L344-L345

which stops the whole parsing at the upper level:

https://github.com/dotnet/runtime/blob/a2ce8db5fb14df85cef03880a30133be198626e1/src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameParser.cs#L103